[Driver] Move CommonArgs to a location visible by the Frontend Drivers

mcinally opened the pull request:
This patch moves the CommonArgs utilities into a location visible by the Frontend Drivers, so that the Frontend Drivers may share option parsing code with the Compiler Driver. This is useful when the Frontend Drivers would like to verify that their incoming options are well-formed and also not reinvent the option parsing wheel.

We already see code in the Clang/Flang Drivers that is parsing and verifying its incoming options. E.g. OPT_ffp_contract. This option is parsed in the Compiler Driver, Clang Driver, and Flang Driver, all with slightly different parsing code. It would be nice if the Frontend Drivers were not required to duplicate this Compiler Driver code. That way there is no/low maintenance burden on keeping all these parsing functions in sync.

Along those lines, the Frontend Drivers will now have a useful mechanism to verify their incoming options are well-formed. Currently, the Frontend Drivers trust that the Compiler Driver is not passing back junk in some cases. The Language Drivers may even accept junk with no error at all. E.g.:

  `clang -cc1 -mprefer-vector-width=junk test.c'

With this patch, we'll now be able to tighten up incomming options to the Frontend drivers in a lightweight way.

Commit:
[Driver] Move CommonArgs to a location visible by the Frontend Drivers

This patch moves the CommonArgs utilities into a location visible by the
Frontend Drivers, so that the Frontend Drivers may share option parsing code
with the Compiler Driver. This is useful when the Frontend Drivers would like
to verify that their incoming options are well-formed and also not reinvent the
option parsing wheel.

We already see code in the Clang/Flang Drivers that is parsing and verifying its
incoming options. E.g. OPT_ffp_contract. This option is parsed in the Compiler
Driver, Clang Driver, and Flang Driver, all with slightly different parsing
code. It would be nice if the Frontend Drivers were not required to duplicate
this Compiler Driver code. That way there is no maintenace burden on keeping all
these parsing functions in sync.

Along those lines, the Frontend Drivers will now have a useful mechanism to
verify their incoming options are well-formed. Currently, the Frontend Drivers
trust that the Compiler Driver is not passing back junk. In some cases, the
Language Drivers will accept junk with no error at all. E.g.:

  `clang -cc1 -mprefer-vector-width=junk test.c'

With this patch, we'll now be able to tighten up incomming options to the
Frontend drivers in a lightweight way.

tarunprabhu commented on clang/include/clang/Driver/CommonArgs.h:
```diff
@@ -270,6 +270,12 @@ void handleVectorizeLoopsArgs(const llvm::opt::ArgList &Args,
 void handleVectorizeSLPArgs(const llvm::opt::ArgList &Args,
                             llvm::opt::ArgStringList &CmdArgs);
+
+void ParseMPreferVectorWidthOption(clang::DiagnosticsEngine &Diags,
```
Is the handling of the mprefer-vector-width option intended to be part of this PR?

tarunprabhu replied:
Ok. Let's leave it in for now. Since this is reorganizing the code in clang, we should wait for the clang developers to weigh in on this.

vzakhari reviewed (approved):
Thank you, Cameron! It looks good to me with one minor comment.

tarunprabhu commented on clang/lib/Driver/ToolChains/CommonArgs.cpp:
```diff
@@ -3167,3 +3167,30 @@ void tools::handleInterchangeLoopsArgs(const ArgList &Args,
+    return;
+  }
+
+  CmdArgs.push_back(Args.MakeArgString(Out + Value));
```
The way this function behaves is consistent with the way options are handled in this part of the code. We could consider renaming the function `handleMPreferVectorWidthOptions` if the word "Parse" suggests that it will return something.

vzakhari replied:
I agree that this aligns with how other functions work. At the same time, it seems awkward to me that we push just the value into the args-list, when the caller is a frontend driver.

mcinally replied:
This is awkward. I was hoping someone with more C++ expertise would have a trick to streamline it.

I chose this solution so that the Compilation Driver side would be as undisturbed as possible.

I also don't feel strongly that it needs to change for an initial implementation. It's something we could build out in the future as we learn more.

tarunprabhu replied:
Argh! Sorry, I somehow missed the real issue here. Sorry, it must have been one of those days.

> That would work. It would disturb the Compiler Driver a bit to avoid warnings. We'd need to explicitly discard the results when unwanted. E.g.:
> 
> `(void)ParseMPreferVectorWidthOption(...);`

Just to clarify, by warnings you mean warnings when building `clang` itself, right?

> Same. I suggest that we delay this decision until there's a better view of the entire problem.

Is the real issue here that we need to validate the argument values? It is a fairly common pattern in, e.g. [`clang/lib/Frontend/CompilerInvocation.cpp`](https://github.com/llvm/llvm-project/blob/main/clang/lib/Frontend/CompilerInvocation.cpp) to look at the `ArgList` and parse the arguments there. It may be better to share the argument validation and return, say, an optional value with the correctly parsed value if it exists. How does that sound?

mcinally replied:
> Just to clarify, by warnings you mean warnings when building clang itself, right?

Correct. The unused result would be a warning.

> Is the real issue here that we need to validate the argument values? It is a fairly common pattern in, e.g. [clang/lib/Frontend/CompilerInvocation.cpp](https://github.com/llvm/llvm-project/blob/main/clang/lib/Frontend/CompilerInvocation.cpp) to look at the ArgList and parse the arguments there.

Yes, that's what we're trying to avoid. Both `Frontend/CompilerInvocation.cpp` and `Driver/FC]lang.cpp` could do verification, but we don't want to have the same/similar code in both places. This patch would let `Frontend/CompilerInvocation.cpp` and `Driver/[FC]lang.cpp` share the same parsing/handling functions.

>  It may be better to share the argument validation and return, say, an optional value with the correctly parsed value if it exists. How does that sound?

I like that!!

I'll wait to see if there are other ideas before updating the PR.

mcinally replied:
Actually, I take that back. This solution is pretty slick. Sharing an update now...

mcinally committed:
Merge branch 'main' into mcinally/junk2

jeanPerier reviewed (approved):
Not familiar with the driver code, but I like the idea of sharing the verification of options.

DavidTruby reviewed (approved):
LGTM thanks for removing the duplication here!

Commit:
[Driver] Move CommonArgs to a location visible by the Frontend Drivers

Update initial patch to use std::optional return values for the parsing
functions. This is a more elegant way of returning just the Values to
the Frontend Drivers.

Commit:
Fix clang-format issues.

shafik commented on clang/lib/Driver/ToolChains/CommonArgs.cpp:
```diff
@@ -3167,3 +3167,29 @@ void tools::handleInterchangeLoopsArgs(const ArgList &Args,
+    return std::nullopt;
+
+  StringRef Value = A->getValue();
+  unsigned Width;
```
Can we add `__attribute((uninitialized))` since this left uninitialized on purpose and is use solely for getAsInteger below.

mcinally replied:
Looks like we can't add this. The MSVC build is failing since `__attribute((uninitialized))` isn't supported there. I'll remove this.

erichkeane replied:
Seems better instead of removing it to add it as a macro in `llvm/include/llvm/Support/Compiler.h`

shafik replied:
Agreed

mcinally committed:
Update clang/lib/Driver/ToolChains/CommonArgs.cpp

Co-authored-by: Shafik Yaghmour <shafik.yaghmour@intel.com>

Commit:
Add __attribute((uninitialized)) to lib/Driver/ToolChains/CommonArgs.cpp

tarunprabhu commented on clang/lib/Driver/ToolChains/CommonArgs.cpp:
```diff
@@ -3167,3 +3167,29 @@ void tools::handleInterchangeLoopsArgs(const ArgList &Args,
+    return std::nullopt;
+  }
+
+  CmdArgs.push_back(Args.MakeArgString(Out + Value));
```
It may be better to always return just the value here, instead of `option + value`. We can still add to `CmdArgs` if `isCompilerDriver=true`. That way, we don't have to make the change in the caller-side. The caller can just ignore the return value. In the case where `isCompilerDriver=false`, we can initialize the options object with the return value as appropriate. Does that sound reasonable?

tarunprabhu replied:
The alternative would be to not add to `CmdArgs` here, but to do it in the caller instead. I think that is ok too. It would be a bit of a departure from the way the rest of the code works, but it feels like a more "natural" way of handling this. The rest of the code could, gradually, be made more consistent with this approach.

mcinally replied:
> The alternative would be to not add to CmdArgs here, but to do it in the caller instead. 

That's a good point as well. There are plenty of `CmdArgs.push_back(...)`'s in `clang/lib/Driver/ToolChains/Clang.cpp` already, so it wouldn't be awkward.

Let me code that up and we can see if we like it...

mcinally replied:
Just pushed a new patch. I like it a lot. See what you think...

We were able to get rid of `isCompilerDriver` and the `clang/lib/Driver/ToolChains/Clang.cpp` change doesn't look awkward at all.

vzakhari replied:
))) this is what I suggested [2 days ago](https://github.com/llvm/llvm-project/pull/142800#discussion_r2127480173)

Thank you, Cameron!

tarunprabhu commented on clang/include/clang/Driver/CommonArgs.h:
```diff
@@ -270,6 +270,11 @@ void handleVectorizeLoopsArgs(const llvm::opt::ArgList &Args,
 void handleVectorizeSLPArgs(const llvm::opt::ArgList &Args,
                             llvm::opt::ArgStringList &CmdArgs);
+
+std::optional<StringRef> ParseMPreferVectorWidthOption(
```
Could we add a docstring for this function

Commit:
Rework ParseMPreferVectorWidthOption(...) to return the Value StringRef.

This patch reworks ParseMPreferVectorWidthOption(...) to only return the Value
StringRef. It will now be the responsibility of the caller to handle the Value
appropriately. This allows more flexibily and makes
ParseMPreferVectorWidthOption(...) useful in different contexts.

Also, add a docstring for this function.

tarunprabhu reviewed (approved):
Thanks for the changes, Cameron. 

Could you take a look at the [failing buildkite](https://github.com/llvm/llvm-project/actions/runs/15493623574/job/43624895590?pr=142800) and see if it is something relevant? It seems to be a Windows build. Those are a bit flaky and a failure doesn't necessarily mean that there is a problem with the PR, but it's probably worth taking a look just in case.

Commit:
Remove __attribute((uninitialized))

Remove __attribute((uninitialized)) as it's not supported on Windows.

mcinally commented:
> Thanks for the changes, Cameron.
> 
> Could you take a look at the [failing buildkite](https://github.com/llvm/llvm-project/actions/runs/15493623574/job/43624895590?pr=142800) and see if it is something relevant? It seems to be a Windows build. Those are a bit flaky and a failure doesn't necessarily mean that there is a problem with the PR, but it's probably worth taking a look just in case.

Thanks! It looks like `__attribute((uninitialized))` isn't supported on Windows, so I have removed it from our patch.

mcinally commented:
It looks like we have a number of approvals and I haven't seen any objections yet.

Should we go ahead with merging this in the current state? If so, I'll need help there since I don't have commit access yet. @tarunprabhu would you mind volunteering to mash the merge button again?

erichkeane commented:
> It looks like we have a number of approvals and I haven't seen any objections yet.
> 
> Should we go ahead with merging this in the current state? If so, I'll need help there since I don't have commit access yet. @tarunprabhu would you mind volunteering to mash the merge button again?

I had 1 minor objection, see my concern on the `uninitialized` attribute removal.

Commit:
Add LLVM_ATTRIBUTE_UNINITIALIZED

Add LLVM_ATTRIBUTE_UNINITIALIZED to map to __attribute__((uninitialized)) when
available.

tarunprabhu commented:
> I had 1 minor objection, see my concern on the `uninitialized` attribute removal.

Would it be simpler just to initialize the variable before use? I assume that the attribute is present to eliminate compiler warnings about uninitialized variables being passed by reference. I don't see the use of this anywhere else in the code base. Does adding that attribute really provide a great deal of meaningful context?

erichkeane reviewed (approved):
Once this passes CI/goes green check, I think we're good to click the 'merge' button.

mcinally commented:
> > I had 1 minor objection, see my concern on the `uninitialized` attribute removal.
> 
> Would it be simpler just to initialize the variable before use? I assume that the attribute is present to eliminate compiler warnings about uninitialized variables being passed by reference. I don't see the use of this anywhere else in the code base. Does adding that attribute really provide a great deal of meaningful context?

This is looking like a better option. Some CIs don't seem to pick up the macro. I'll do some investigating, but I'm not able to reproduce with my build.

Commit:
Fix typo in LLVM_ATTRIBUTE_UNINITIALIZED.

Missing 'Z'.

tarunprabhu commented on llvm/include/llvm/Support/Compiler.h:
```diff
@@ -238,6 +238,12 @@
 #endif
 
+#if __has_attribute(uninitialized)
+#define LLVM_ATTRIBUTE_UNINITIALIED __attribute__((uninitialized))
```
Spelling: 

```suggestion
#define LLVM_ATTRIBUTE_UNINITIALIZED __attribute__((uninitialized))
```

mcinally committed:
Merge branch 'main' into mcinally/junk2

eugeneepshteyn commented on clang/lib/Driver/ToolChains/CommonArgs.cpp:
```diff
@@ -3167,3 +3167,25 @@ void tools::handleInterchangeLoopsArgs(const ArgList &Args,
+    return "";
+
+  StringRef Value = A->getValue();
+  unsigned Width LLVM_ATTRIBUTE_UNINITIALIZED;
```
This is fine, but FWIW, there's C++ `[[maybe_unused]]` that could possibly work here: https://en.cppreference.com/w/cpp/language/attributes/maybe_unused
